File: go/modcdp/transport/WebSocketUpstreamTransport.go
package transport

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"sync"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

type WebSocketUpstreamTransport struct {
	UpstreamTransport
	URL     string
	Conn    net.Conn
	writeMu sync.Mutex
}

type WebSocketUpstreamTransportOptions struct {
	CDPURL string `json:"cdp_url,omitempty"`
}

func NewWebSocketUpstreamTransport(options WebSocketUpstreamTransportOptions) *WebSocketUpstreamTransport {
	return &WebSocketUpstreamTransport{URL: options.CDPURL}
}

func (t *WebSocketUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	if value, ok := config["cdp_url"].(string); ok && value != "" {
		t.URL = value
	}
}

func (t *WebSocketUpstreamTransport) Connect() error {
	if t.URL == "" {
		return fmt.Errorf("WebSocketUpstreamTransport requires upstream_cdp_url or launcher-provided cdp_url")
	}
	// URL may start as an HTTP cdp_url; from here on it is the resolved WebSocket CDP endpoint.
	resolvedURL, err := websocketURLFor(t.URL)
	if err != nil {
		return err
	}
	t.URL = resolvedURL
	conn, _, _, err := ws.Dial(context.Background(), t.URL)
	if err != nil {
		return err
	}
	t.writeMu.Lock()
	t.Conn = conn
	t.writeMu.Unlock()
	go t.readLoop(conn)
	return nil
}

func (t *WebSocketUpstreamTransport) Send(message map[string]any) error {
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	conn := t.Conn
	if conn == nil {
		return fmt.Errorf("CDP websocket is not connected")
	}
	return wsutil.WriteClientText(conn, body)
}

func (t *WebSocketUpstreamTransport) Close() error {
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	if t.Conn != nil {
		err := t.Conn.Close()
		t.Conn = nil
		return err
	}
	return nil
}

func (t *WebSocketUpstreamTransport) readLoop(conn net.Conn) {
	for {
		data, err := wsutil.ReadServerText(conn)
		if err != nil {
			t.writeMu.Lock()
			currentConn := t.Conn
			t.writeMu.Unlock()
			if currentConn == conn {
				t.emitClose(err)
			}
			return
		}
		var message map[string]any
		if err := json.Unmarshal(data, &message); err == nil {
			t.emitRecv(message)
		}
	}
}

File: js/src/transport/WebSocketUpstreamTransport.ts
import { resolveCdpWebSocketUrl } from "../launcher/BrowserLauncher.js";
import { UpstreamTransport, type UpstreamTransportConfig } from "./UpstreamTransport.js";

export class WebSocketUpstreamTransport extends UpstreamTransport {
  readonly upstream_mode = "ws" as const;
  readonly endpoint_kind = "raw_cdp" as const;
  ws: WebSocket | null = null;

  constructor({ cdp_url = null }: { cdp_url?: string | null } = {}) {
    super();
    this.upstream_cdp_url = cdp_url ?? "";
    this.send_command = (message) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) throw new Error("CDP websocket is not connected.");
      this.ws.send(JSON.stringify(message));
    };
  }

  update(config: UpstreamTransportConfig = {}) {
    if (config.cdp_url) this.upstream_cdp_url = config.cdp_url;
    if (typeof config.cdp_send_timeout_ms === "number") this.cdp_send_timeout_ms = config.cdp_send_timeout_ms;
    return this;
  }

  getServerConfig() {
    return this.upstream_cdp_url ? { server_loopback_cdp_url: this.upstream_cdp_url } : {};
  }

  async connect() {
    if (!this.upstream_cdp_url)
      throw new Error("upstream.upstream_mode=ws requires upstream_cdp_url or launcher-provided cdp_url.");
    // cdp_url may start as an HTTP discovery endpoint; from here on it is the resolved WebSocket CDP endpoint.
    this.upstream_cdp_url = await resolveCdpWebSocketUrl(this.upstream_cdp_url, "upstream_cdp_url");
    const ws = new WebSocket(this.upstream_cdp_url);
    this.ws = ws;
    ws.addEventListener("message", (event) => this.parseAndEmitRecv(event.data));
    ws.addEventListener("close", () => this.emitClose(new Error("CDP websocket closed")));
    ws.addEventListener("error", () => this.emitClose(new Error("CDP websocket error")));
    await new Promise<void>((resolve, reject) => {
      const cleanup = () => {
        ws.removeEventListener("open", onOpen);
        ws.removeEventListener("error", onError);
      };
      const onOpen = () => {
        cleanup();
        resolve();
      };
      const onError = () => {
        cleanup();
        reject(new Error("CDP websocket error"));
      };
      ws.addEventListener("open", onOpen);
      ws.addEventListener("error", onError);
    });
  }

  async close() {
    try {
      this.ws?.close();
    } catch {}
    this.ws = null;
  }
}
